### PR TITLE
Add ability to sort friends by rank

### PR DESCRIPTION
**Pull request by venix12**
Adds an option to sort friends by default gamemode rank 
TODO: make eager-load work properly

fixes https://github.com/ppy/osu-web/issues/4759

### CODE REVIEW

File: app/Transformers/UserCompactTransformer.php [hunk 1]
```diff
@@ -48,6 +48,7 @@ public function transform(User $user)
             'is_supporter' => $user->isSupporter(),
             'last_visit' => json_time($user->displayed_last_visit),
             'pm_friends_only' => $user->pm_friends_only,
+            'user_gamemode_rank' => $user->statistics($user->getPlaymodeAttribute($user->osu_playmode), 'true')->rank,
```
Comment (notbakaneko) [inline app/Transformers/UserCompactTransformer.php]: `statistics` can return `null`

Comment (notbakaneko) [reply app/Transformers/UserCompactTransformer.php]: `$user->playmode`

Comment (notbakaneko) [reply app/Transformers/UserCompactTransformer.php]: `'true'` is a string not boolean; also this ends up being `false` in the statistics conditional

Comment (notbakaneko) [reply app/Transformers/UserCompactTransformer.php]: this ends up being the rank for the user's selected default game mode, so the property name sounds misleading.

File: resources/assets/lib/user-list.tsx [hunk 2]
```diff
@@ -44,6 +44,10 @@ function usernameSortAscending(x: User, y: User) {
   return x.username.localeCompare(y.username);
 }
 
+function rankSortDescending(x: User, y: User) {
+ return x.user_gamemode_rank > y.user_gamemode_rank;
```
Comment (notbakaneko) [inline resources/assets/lib/user-list.tsx]: properties added to the type should be added to `globals.d.ts`

Comment (notbakaneko) [reply resources/assets/lib/user-list.tsx]: `compareFunction` for `Array.sort` should return a `number`
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/sort

File: resources/lang/en/sort.php [hunk 3]
```diff
@@ -26,6 +26,7 @@
     'old' => 'Old',
     'top' => 'Top',
     'username' => 'Username',
+    'rank' => 'Rank',
```
Comment (notbakaneko) [inline resources/lang/en/sort.php]: keys should be alphabetical

File: resources/assets/lib/user-list.tsx [hunk 4]
```diff
@@ -63,6 +67,8 @@ export class UserList extends React.PureComponent<Props> {
     switch (this.state.sortMode) {
       case 'username':
         return users.sort(usernameSortAscending);
+      case 'rank':
```
Comment (notbakaneko) [inline resources/assets/lib/user-list.tsx]: case rank before username

**Comment by VenixOSU** [comment]
Will work on eager-load soon

**Comment by venix12** [comment]
would be better if i do this in a new pr i believe

**Pull request closed without merging**